pwndbg/aglib/arch.py

```diff
@@ -1,39 +1,319 @@
+        return (CS_ARCH_X86, CS_MODE_16)
+
+
+class ArmArch(PwndbgArchitecture):
+    def __init__(self) -> None:
+        super().__init__("arm")
+
+    @override
+    def get_capstone_constants(self, address: int) -> Tuple[int, int]:
+        thumb_mode = pwndbg.aglib.disasm.emulated_arm_mode_cache[address]
+        if thumb_mode is None:
+            thumb_mode = self.read_thumb_bit()
+        mode = CS_MODE_THUMB if thumb_mode else CS_MODE_ARM
+
+        return (CS_ARCH_ARM, mode)
+
+    @override
+    def read_thumb_bit(self) -> Literal[0, 1]:
         # When program initially starts, cpsr may not be readable
         if (cpsr := pwndbg.aglib.regs.cpsr) is not None:
-            return (cpsr >> 5) & 1
-    elif pwndbg.aglib.arch.name == "armcm":
-        # ARM Cortex-M procesors only suport Thumb mode. However, there is still a bit
-        # that represents the Thumb mode (which is currently architecturally defined to be 1)
-        if (xpsr := pwndbg.aglib.regs.xpsr) is not None:
-            return (xpsr >> 24) & 1
-    # AArch64 does not have a Thumb bit
-    return None
+            return (cpsr >> 5) & 1  # type: ignore[return-value]
 
+        return 0
 
-def get_thumb_mode_string() -> Literal["arm", "thumb"] | None:
-    thumb_bit = read_thumb_bit()
-    return None if thumb_bit is None else "thumb" if thumb_bit == 1 else "arm"
+
+class ArmCortexArch(PwndbgArchitecture):
+    """
+    Cortex-M processors run the M-profile Arm architecture.
+    This architecture is prevalent in bare-metal/embedded systems that lack operating systems.
+    Only Thumb-2 instructions are supported, and the Thumb bit is always 1.
+    """
+
+    def __init__(self) -> None:
+        super().__init__("armcm")
+
+    @override
+    def get_capstone_constants(self, address: int) -> Tuple[int, int]:
+        return (CS_ARCH_ARM, CS_MODE_MCLASS | CS_MODE_THUMB)
+
+    @override
+    def read_thumb_bit(self) -> Literal[0, 1]:
+        """
+        On Cortex-M processors, the Thumb bit is architecturally defined to be 1.
+
+        This is the (xpsr >> 24) & 1, which is always 1.
+        """
+        return 1
+
+
+class AArch64Arch(PwndbgArchitecture):
+    def __init__(self) -> None:
+        super().__init__("aarch64")
+
+    def get_capstone_constants(self, address: int) -> Tuple[int, int]:
+        return (CS_ARCH_ARM64, CS_MODE_ARM)
+
+
+class PowerPCArch(PwndbgArchitecture):
+    def __init__(self) -> None:
+        super().__init__("powerpc")
+
+    def get_capstone_constants(self, address: int) -> Tuple[int, int]:
+        # TODO - what about 32-bit powerpc?
+        return (CS_ARCH_PPC, CS_MODE_64)
 
 
-arch: Arch = Arch("i386", 4, "little")
+class SparcArch(PwndbgArchitecture):
+    def __init__(self) -> None:
+        super().__init__("sparc")
+
+    def get_capstone_constants(self, address: int) -> Tuple[int, int]:
+        if pwndbg.dbg.is_gdblib_available() and "v9" in self.name_raw:
```

Comment:
```suggestion
        if "v9" in self.name_raw:
```

`pwndbg.dbg.is_gdblib_available()` is not needed here after removing gdb